### PR TITLE
Improving code to search all userstores so userstore domain in the certificate is not required

### LINKED ISSUE #6762: X509 Authenticator authenticates users only in the primary userstore.
The current implementation checks for the user specified in the CN or Subject Alternative Name in the primary user store, if the user is in a secondary store the user will have to be defined in the following format in the certificate: USER-STORE-NAME/USER-NAME.

### PR DESCRIPTION
**Pull request by tharakawijekoon**
Resolves https://github.com/wso2/product-is/issues/6762
This improvement will search for the username specified in the certificate in all user stores, if there are conflicting users with the same username across multiple userstores, authentication will fail.

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateConstants.java [hunk 1]
```diff
@@ -67,5 +67,6 @@ private X509CertificateConstants() {
             + "names in the certificate";
     public static final String X509_CERTIFICATE_SUBJECTDN_REGEX_NO_MATCHES_ERROR = "Regex configured but no matching "
             + "subjectRDN found for the given regex";
+    public static final int MAX_ITEM_LIMIT_UNLIMITED = -1;
 
 }
```
Comment (janakamarasena) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateConstants.java]: Add empty line EOF

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java [hunk 2]
```diff
@@ -323,13 +323,17 @@ private static boolean isUserCertificateValid(String userName, X509Certificate x
     private static boolean isUserExists(String userName, AuthenticationContext authenticationContext)
             throws UserStoreException, AuthenticationFailedException {
 
-        boolean isUserExist = X509CertificateUtil.getUserRealm(userName).getUserStoreManager().isExistingUser
-                (MultitenantUtils.getTenantAwareUsername(userName));
-        if (isUserExist) {
+        String[] filteredUsers = X509CertificateUtil.getUserRealm(userName).getUserStoreManager().listUsers
+                (MultitenantUtils.getTenantAwareUsername(userName),X509CertificateConstants.MAX_ITEM_LIMIT_UNLIMITED);
+        if(filteredUsers.length==1){
```
Comment (janakamarasena) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: Fix code formatting

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java [hunk 3]
```diff
@@ -323,13 +323,17 @@ private static boolean isUserCertificateValid(String userName, X509Certificate x
...
             if (log.isDebugEnabled()) {
                 log.debug("User exists with the user name: " + userName);
             }
             return true;
+        } else if (filteredUsers.length>1){
+            authenticationContext.setProperty(X509CertificateConstants.X509_CERTIFICATE_ERROR_CODE,
+                    X509CertificateConstants.USERNAME_CONFLICT);
+            throw new AuthenticationFailedException(" Conflicting users in user stores. ");
```
Comment (janakamarasena) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: Fix formatting. Shall we add the user details in the exception message?

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java [hunk 4]
```diff
@@ -323,13 +323,17 @@ private static boolean isUserCertificateValid(String userName, X509Certificate x
...
+                (MultitenantUtils.getTenantAwareUsername(userName), X509CertificateConstants.MAX_ITEM_LIMIT_UNLIMITED);
+        if (filteredUsers.length == 1) {
             if (log.isDebugEnabled()) {
                 log.debug("User exists with the user name: " + userName);
             }
             return true;
+        } else if (filteredUsers.length > 1) {
+            authenticationContext.setProperty(X509CertificateConstants.X509_CERTIFICATE_ERROR_CODE,
+                    X509CertificateConstants.USERNAME_CONFLICT);
+            throw new AuthenticationFailedException(" Conflicting users with user name: " + userName);
```
Comment (Buddhimah) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: Shall we remove the space here : ("Conflicting users with user name: " + userName);

**Review by Buddhimah — APPROVED** [review]

**Comment by janakamarasena** [comment]
@tharakawijekoon, since this will introduce a behavioural change lets, in order to preserve backward compatibility add a property to enable this functionality. You can add the property to the application-authentication.xml file under the relevant authenticator.

Ideally 

- If the domain qualifier is there in the CN then the user should be picked from the particular user store.
- Otherwise if a property is enabled, it will iterate through all the userstores. If the same user name conflicts among multiple userstores the authentication will be failed.

**Review by janakamarasena — CHANGES_REQUESTED** [review]
@tharakawijekoon, since this will introduce a behavioural change, in order to preserve backward compatibility lets add a property to enable this functionality. You can add the property to the application-authentication.xml file under the relevant authenticator.

Ideally 

- If the domain qualifier is there in the CN then the user should be picked from the particular user store.
- Otherwise if a property is enabled, it will iterate through all the userstores. If the same user name conflicts among multiple userstores the authentication will be failed.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java [hunk 5]
```diff
@@ -138,8 +138,8 @@ public static boolean addCertificate(String username, X509Certificate x509Certif
      * @throws AuthenticationFailedException
      */
     public static boolean validateCertificate(String userName, AuthenticationContext authenticationContext,
-                                              byte[] certificateBytes, boolean isSelfRegistrationEnable)
-            throws AuthenticationFailedException {
+                                              byte[] certificateBytes, boolean isSelfRegistrationEnable,
+                                              boolean searchSecondaryUserStores) throws AuthenticationFailedException {
```
Comment (DMHP) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: We can't change the method signatures of the existing public methods . There can be external codes which rely on this public method of this Util. Hence we need to deprecate the existing method and introduce a new method with the new signature.

Comment (tharakawijekoon) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/x509Certificate/X509CertificateUtil.java]: I called the getX509Parameters() to get the parameter which is defined in the application-authentication.xml.
The "SearchAllUserStores" parameter should be defined as follows:
```<AuthenticatorConfig name="x509CertificateAuthenticator" enabled="true">
            <Parameter name="AuthenticationEndpoint">https://localhost:443/x509-certificate-servlet</Parameter>
            <!-- <Parameter name="username">CN</Parameter> -->
            <!-- <Parameter name="UsernameRegex">[a-zA-Z]{3}</Parameter> -->
            <Parameter name="AlternativeNamesRegex">[a-zA-Z]{7}</Parameter>
            <Parameter name="SearchAllUserStores">true</Parameter>
            <!-- <Parameter name="X509UserStoreName">SECONDARY</Parameter> -->
            <!--<Parameter name="setClaimURI">http://wso2.org/claims/identity/userCertificate</Parameter>-->
            <!--<Parameter name="EnforceSelfRegistration">true</Parameter>-->
        </AuthenticatorConfig>

**Review by janakamarasena — APPROVED** [review]

**Review by nilasini — APPROVED** [review]

**Pull request merged**